### ISSUE TITLE
systemd-networkd does not allow for the same IP to be statically assigned to multiple MAC addresses

### ISSUE THREAD

**Issue by vojta001** [systemd/systemd#27720]
### systemd version the issue has been seen with

253.3-3

### Used distribution

Archlinux

### Linux kernel version used

6.2.12-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

networkd loads the configuration and then assigns the same IP address to all the MAC addresses specified.

Having such a feature is surprisingly useful. By assigning the same IP to both your wired and wireless interface, you can hop from one to the other without interrupting existing TCP connections.

### Unexpected behaviour you saw

After `networkctl reload`, `networkctl status` says

```
br0: Failed to set DHCPv4 static lease for DHCP server: File exists
```

### Steps to reproduce the problem

Create a network interface (e. g. a bridge) with DHCP server turned on and multiple static leases for different MAC addresses assigning the same IP address

e. g.

```
[Match]
Name=br0

[Network]
DHCPServer=yes
Address=10.1.2.1/24
LinkLocalAddressing=no
IPv6AcceptRA=no
IPv6SendRA=no

[DHCPServer]
PoolOffset=100
PoolSize=120
DefaultLeaseTimeSec=12h
MaxLeaseTimeSec=24h
DNS=1.1.1.1
Router=10.1.2.1
EmitTimezone=yes
BindToInterface=yes


[DHCPServerStaticLease]
MACAddress=aa:bb:cc:dd:ee:ff
Address=10.1.2.2

[DHCPServerStaticLease]
MACAddress=11:22:33:44:55:66
Address=10.1.2.2
```

**Comment by ssahani** [comment]
This sounds more like ```bonding (active-backup)```. It  could result a IP conflict if added same IP twice by mistake.

**Comment by vojta001** [comment]
Configuring your client to bridge Wi-Fi and ethernet together and then backup-bonding them is very complex and depends on you having permission to make such deep configuration changes (it can be e. g. a work laptop).

Configuring the DHCP server to assign the same IP address is on the other hand very simple, reliable and doesn't require any configuration changes on the client.

**Comment by ssahani** [comment]
[This RFE actually defeats the purpose of ](https://github.com/systemd/systemd/issues/22830) Here is An example. 

>Configuring the DHCP server to assign the same IP address is on the other hand very simple, reliable and doesn't require any configuration changes on the client.

This actually defeats the whole purpose of DHCP i.e unique address.

**Comment by vojta001** [comment]
> (https://github.com/systemd/systemd/issues/22830) Here is An example.

Yes, but it doesn't solve the problem: You must have admin rights on the client.

> This actually defeats the whole purpose of DHCP i.e unique address.

I don't think so. Having unique active addresses on a network and having unique addresses in DHCP configuration are two completely orthogonal things.

What I want is to come home, connect to WiFi, use internet, then when in need of a faster connection put my laptop on a desk, connect cable, disconnect WiFi, profit. And I want all of these to happen without trashing TCP connection, which implies getting the same IP on the cable.

**Comment by vojta001** [comment]
And yes, if you forget to disable one of the interfaces, you get an IP conflict and very suboptimal speeds (things will still work though. Your ifs will just constantly compete for switch's forwarding tables).

I just don't thing networkd should actively block this valid usecase. A warning or a `YesIKnowWhatIAmDoing` which would be fine.

**Comment by rcambrj** [comment]
I'm so glad that you raised this issue @vojta001 - it was the one and only result when searching for:

```
Failed to set DHCPv4 static lease for DHCP server: File exists
```

In my case, I have a machines with multiple NICs, each with their own hardware MAC address, plus one which is randomly generated for the bridge. I don't want to care where which port I connect this machine with - it should always get the same IP address. When the bridge is configured, this is fine, but sometimes the bridge is not configured, and that's the problematic case with systemd-networkd. The bridge is sometimes not configured while I'm setting up these machines with a new configuration.

Coming from `dnsmasq` where multiple hardware MAC addresses **are** permitted per IP address configuration, I was confused by this error message. `dnsmasq` almost certainly warns/errors/panics at the point when the static lease is requested by >1 hardware MAC addresses, but I've never had that happen, so I can't confirm.

**Comment by blsz** [comment]
The scenario described by @vojta001 is very useful for making SSH connections to a laptop behind a router using the laptop's name or IP address without knowing which interface - wired or wireless - is active.  It is very common to assign the same IP address to multiple MAC addresses so that the device can be addressed by a consistent name or IP.    Otherwise extra steps are needed to determine the IP address of the device.  If the device has 2 different IP's, then how do you know which is active and which to SSH to?

This is not contrary to any principles of DHCP.  It works without any issues as long as the client does not try to bring up both interfaces at the same time. 

```
% man dnsmasq
       -G, --dhcp-host=[<hwaddr>][,id:<client_id>|*][,set:<tag>][,<ipaddr>][,<hostname>][,<lease_time>][,ignore]
...
              As  a  special  case,  in  DHCPv4,  it  is  possible  to  include  more  than  one  hardware  address. eg: --dhcp-
              host=11:22:33:44:55:66,12:34:56:78:90:12,192.168.0.2 This allows an IP address  to  be  associated  with  multiple
              hardware  addresses,  and  gives  dnsmasq permission to abandon a DHCP lease to one of the hardware addresses when
              another one asks for a lease. Beware that this is a dangerous thing to do, it will only work reliably if only  one
              of  the  hardware  addresses  is  active  at  any time and there is no way for dnsmasq to enforce this. It is, for
              instance, useful to allocate a stable IP address to a laptop which has both wired and wireless interfaces.
```

Please consider making this an allowed option in systemd-networkd.  Even though this configuration may be considered "dangerous",  it is extremely useful and often relied on.   It would be very useful to have an option that allowed the same IP to be assigned dynamically, as well as statically, to multiple MAC addresses while preventing networkd from bringing up interfaces with the same IP address at the same time.
